File: hw_6/D.py
from typing import Tuple, List
from random import randint
import sys


class ImplicitTreap:
    def __init__(self, x: int, left_treap=None, right_treap=None):
        self.x: int = x
        self.y: int = randint(1, 150000)
        self.left: ImplicitTreap = left_treap
        self.right: ImplicitTreap = right_treap

        self.size: int = 1
        self.reversed = False


def merge(L: ImplicitTreap, R: ImplicitTreap) -> ImplicitTreap:
    push(L)
    push(R)
    if L is None:
        return R
    if R is None:
        return L

    if L.y > R.y:
        L.right = merge(L.right, R)
        recalc(L)
        return L
    else:
        R.left = merge(L, R.left)
        recalc(R)
        return R


def split(T: ImplicitTreap, x0: int) -> Tuple[ImplicitTreap, ImplicitTreap]:
    push(T)
    if T is None:
        return (None, None)

    if get_size(T.left) < x0:
        L, R = split(T.right, x0 - get_size(T.left) - 1)
        T.right = L
        recalc(T)
        return (T, R)
    else:
        L, R = split(T.left, x0)
        T.left = R
        recalc(T)
        return (L, T)


def get_size(T: ImplicitTreap):
    if T is None:
        return 0
    else:
        return T.size


def recalc(T: ImplicitTreap):
    if T is not None:
        T.size = get_size(T.left) + get_size(T.right) + 1


def push(T: ImplicitTreap):
    if T is None:
        return
    if T.reversed is False:
        return

    T.reversed = False
    T.left, T.right = T.right, T.left
    if T.left is not None:
        T.left.reversed ^= True
    if T.right is not None:
        T.right.reversed ^= True


def reverse(T: ImplicitTreap, A: int, B: int) -> ImplicitTreap:
    # Split by left if A = 3, B = 7 <=> [1, 2, 3, 4, 5, 6, 7, 8] -> [1, 2, 3] & [4, 5, 6, 7, 8]
    L, R = split(T, A - 1)

    # Split by right [4, 5, 6, 7, 8, 9] -> [4, 5, 6, 7] & [8]
    M, R = split(R, B - A + 1)
    M.reversed ^= True

    return merge(merge(L, M), R)


def print_tree(T: ImplicitTreap):
    if T is None:
        return
    push(T)
    print_tree(T.left)
    print(T.x, end=' ')
    print_tree(T.right)


def from_list(arr: List[int]) -> ImplicitTreap:
    result = None
    for val in arr:
        result = merge(result, ImplicitTreap(val))
    return result


if __name__ == "__main__":
    n, m = [int(i) for i in input().split(' ')]
    arr = [i for i in range(1, n+1)]

    main_T = from_list(arr)
    # print_tree(main_T)
    # print()

    while m:
        a, b = map(int, sys.stdin.readline().split(' '))
        reverse(main_T, a, b)
        # print_tree(main_T)
        # print()
        m -= 1

    print_tree(main_T)

File: hw_6/D.cpp
#include <iostream>
#include <vector>
#include <random>
#include <ctime>

struct ImplicitTreap {
    int x;
    int y;
    ImplicitTreap* left;
    ImplicitTreap* right;
    int size;
    bool reversed;

    ImplicitTreap(int x, ImplicitTreap* left = nullptr, ImplicitTreap* right = nullptr) {
        this->x = x;
        this->y = rand();
        this->left = left;
        this->right = right;
        this->size = 1;
        this->reversed = false;
    }
};

int get_size(ImplicitTreap* T) {
    if (T == nullptr) {
        return 0;
    } else {
        return T->size;
    }
}

void recalc(ImplicitTreap* T) {
    if (T != nullptr) {
        T->size = get_size(T->left) + get_size(T->right) + 1;
    }
}

void push(ImplicitTreap* T) {
    if (T == nullptr) {
        return;
    }
    if (!T->reversed) {
        return;
    }
    T->reversed = false;
    std::swap(T->left, T->right);
    if (T->left != nullptr) {
        T->left->reversed ^= true;
    }
    if (T->right != nullptr) {
        T->right->reversed ^= true;
    }
}

ImplicitTreap* merge(ImplicitTreap* L, ImplicitTreap* R) {
    push(L);
    push(R);
    if (L == nullptr) {
        return R;
    }
    if (R == nullptr) {
        return L;
    }
    if (L->y > R->y) {
        L->right = merge(L->right, R);
        recalc(L);
        return L;
    } else {
        R->left = merge(L, R->left);
        recalc(R);
        return R;
    }
}

std::pair<ImplicitTreap*, ImplicitTreap*> split(ImplicitTreap* T, int x0) {
    push(T);
    if (T == nullptr) {
        return {nullptr, nullptr};
    }
    if (get_size(T->left) < x0) {
        auto [L, R] = split(T->right, x0 - get_size(T->left) - 1);
        T->right = L;
        recalc(T);
        return {T, R};
    } else {
        auto [L, R] = split(T->left, x0);
        T->left = R;
        recalc(T);
        return {L, T};
    }
}

ImplicitTreap* reverse(ImplicitTreap* T, int A, int B) {
    auto [L, R] = split(T, A - 1);
    auto [M, R2] = split(R, B - A + 1);
    M->reversed ^= true;
    return merge(merge(L, M), R2);
}

void print_tree(ImplicitTreap* T) {
    if (T == nullptr) {
        return;
    }
    push(T);
    print_tree(T->left);
    std::cout << T->x << " ";
    print_tree(T->right);
}

ImplicitTreap* from_list(const std::vector<int>& arr) {
    ImplicitTreap* result = nullptr;
    for (int val : arr) {
        result = merge(result, new ImplicitTreap(val));
    }
    return result;
}

int main() {
    srand(time(nullptr));

    int n, m;
    std::cin >> n >> m;

    std::vector<int> arr(n);
    for (int i = 0; i < n; ++i) {
        arr[i] = i + 1;
    }

    ImplicitTreap* main_T = from_list(arr);

    while (m--) {
        int a, b;
        std::cin >> a >> b;
        main_T = reverse(main_T, a, b);
    }

    print_tree(main_T);

    return 0;
}